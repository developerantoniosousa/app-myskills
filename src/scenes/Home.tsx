import React, { useState, useEffect } from 'react';
import {
  View,
  Text,
  StyleSheet,
  TextInput,
  Platform,
  FlatList,
} from 'react-native';

import { Button } from './components/Button';
import { SkillCard } from './components/SkillCard';

export interface Skill {
  id: string;
  name: string;
}

export enum GREETING {
  GOOD_MORNING = 'Good morning',
  GOOD_AFTERNOON = 'Good afternoon',
  GOOD_NIGHT = 'Good night',
}

export function Home() {
  const [newSkill, setNewSkill] = useState('');
  const [mySkills, setMySkills] = useState<Skill[]>([]);
  const [greeting, setGreeting] = useState('');

  useEffect(() => {
    const currentHour = new Date().getHours();

    if (currentHour < 12) {
      setGreeting(GREETING.GOOD_MORNING);
    } else if (currentHour >= 12 && currentHour < 18) {
      setGreeting(GREETING.GOOD_AFTERNOON);
    } else {
      setGreeting(GREETING.GOOD_NIGHT);
    }
  }, []);

  function handleAddNewSkill() {
    if (!newSkill) {
      return;
    }
    const addingSkill = {
      id: String(Date.now()),
      name: newSkill,
    };
    setMySkills(existedSkills => [...existedSkills, addingSkill]);
    setNewSkill('');
  }

  function handleRemoveSkill(id: string) {
    setMySkills(existedSkills =>
      existedSkills.filter(skill => skill.id !== id),
    );
  }

  return (
    <View style={styles.container}>
      <Text style={styles.title}>Welcome, Antonio</Text>
      <Text style={styles.greeting}>{greeting}</Text>

      <TextInput
        style={styles.input}
        placeholder="New skill"
        placeholderTextColor="#555"
        value={newSkill}
        onChangeText={setNewSkill}
      />
      <Button title="Add" activeOpacity={0.7} onPress={handleAddNewSkill} />

      <Text style={[styles.title, styles.mySkills]}>My Skills</Text>
      <FlatList
        data={mySkills}
        keyExtractor={item => item.id}
        renderItem={({ item: skill }) => (
          <SkillCard
            key={skill.id}
            skill={skill.name}
            onLongPress={() => handleRemoveSkill(skill.id)}
          />
        )}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#121015',
    paddingVertical: 70,
    paddingHorizontal: 30,
  },
  title: {
    color: '#fff',
    fontSize: 24,
    fontWeight: 'bold',
  },
  input: {
    backgroundColor: '#1F1E25',
    color: '#fff',
    fontSize: 18,
    padding: Platform.OS === 'ios' ? 15 : 10,
    marginTop: 30,
    borderRadius: 7,
  },
  mySkills: {
    marginVertical: 50,
  },
  greeting: {
    color: '#FFF',
  },
});
