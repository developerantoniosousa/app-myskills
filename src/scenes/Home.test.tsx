import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';

import { Home, GREETING } from './Home';

enum DAY_PERIODS {
  MORNING = 'morning',
  AFTERNOON = 'afternoon',
  NIGHT = 'night',
}

function setDayPeriod(period: DAY_PERIODS): void {
  if (period === DAY_PERIODS.MORNING) {
    global.Date.prototype.getHours = function () {
      return 8;
    };
  }

  if (period === DAY_PERIODS.AFTERNOON) {
    global.Date.prototype.getHours = function () {
      return 14;
    };
  }

  if (period === DAY_PERIODS.NIGHT) {
    global.Date.prototype.getHours = function () {
      return 21;
    };
  }
}

describe('<Home />', () => {
  const skillText = 'Javascript';

  test('render default <Home /> without performing any action', () => {
    setDayPeriod(DAY_PERIODS.NIGHT);
    const { toJSON } = render(<Home />);
    expect(toJSON()).toMatchSnapshot();
  });

  test('render <Home /> at morning', () => {
    setDayPeriod(DAY_PERIODS.MORNING);
    const { getByText } = render(<Home />);
    getByText(GREETING.GOOD_MORNING);
  });

  test('render <Home /> at afternoon', () => {
    setDayPeriod(DAY_PERIODS.AFTERNOON);
    const { getByText } = render(<Home />);
    getByText(GREETING.GOOD_AFTERNOON);
  });

  test('render <Home /> at night', () => {
    setDayPeriod(DAY_PERIODS.NIGHT);
    const { getByText } = render(<Home />);
    getByText(GREETING.GOOD_NIGHT);
  });

  test('add a new skill on the skills list', () => {
    const { getByPlaceholderText, getByText } = render(<Home />);

    const inputElement = getByPlaceholderText('New skill');
    const buttonElement = getByText('Add');

    fireEvent.changeText(inputElement, skillText);
    fireEvent.press(buttonElement);

    getByText(skillText);
  });

  test("don't put skill on the list if user does not type anything", async () => {
    const { getByText, queryByText } = render(<Home />);

    const buttonElement = getByText('Add');
    fireEvent.press(buttonElement);

    expect(await queryByText('')).toBeNull();
  });

  test('remove skill from list', async () => {
    const { getByText, getByPlaceholderText, queryByText } = render(<Home />);

    const inputElement = getByPlaceholderText('New skill');
    const buttonElement = getByText('Add');

    fireEvent.changeText(inputElement, skillText);
    fireEvent.press(buttonElement);

    const addedSkillElement = getByText(skillText);
    expect(addedSkillElement).toBeDefined();

    fireEvent(addedSkillElement, 'onLongPress');
    expect(await queryByText(skillText)).toBeNull();
  });
});
