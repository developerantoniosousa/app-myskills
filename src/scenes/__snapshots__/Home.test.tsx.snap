// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> render default <Home /> without performing any action 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#121015",
      "flex": 1,
      "paddingHorizontal": 30,
      "paddingVertical": 70,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#fff",
        "fontSize": 24,
        "fontWeight": "bold",
      }
    }
  >
    Welcome, Antonio
  </Text>
  <Text
    style={
      Object {
        "color": "#FFF",
      }
    }
  >
    Good night
  </Text>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="New skill"
    placeholderTextColor="#555"
    rejectResponderTermination={true}
    style={
      Object {
        "backgroundColor": "#1F1E25",
        "borderRadius": 7,
        "color": "#fff",
        "fontSize": 18,
        "marginTop": 30,
        "padding": 15,
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#A370F7",
        "borderRadius": 7,
        "marginTop": 20,
        "opacity": 1,
        "padding": 15,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFF",
          "fontSize": 17,
          "fontWeight": "bold",
        }
      }
    >
      Add
    </Text>
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "#fff",
          "fontSize": 24,
          "fontWeight": "bold",
        },
        Object {
          "marginVertical": 50,
        },
      ]
    }
  >
    My Skills
  </Text>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</View>
`;
